experimental/dds/tree2/src/core/tree/treeIndex.ts

```diff
@@ -0,0 +1,161 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/core-utils";
+import {
+	Brand,
+	IdAllocator,
+	NestedMap,
+	brand,
+	decodeNestedMap,
+	deleteFromNestedMap,
+	encodeNestedMap,
+	idAllocatorFromMaxId,
+	populateNestedMap,
+	setInNestedMap,
+	tryGetFromNestedMap,
+} from "../../util";
+import { FieldKey } from "../schema-stored";
+import * as Delta from "./delta";
+
+/**
+ * ID used to create a detached field key for a removed subtree.
+ * @alpha
+ *
+ * TODO: Move to Forest once forests can support multiple roots.
+ */
+export type ForestRootId = Brand<number, "tree.ForestRootId">;
+
+export interface Entry {
+	field: FieldKey;
+	root: ForestRootId;
+}
+
+type Major = string | number | undefined;
+type Minor = number;
+
+/**
+ * The tree index records detached field ids and associates them with a change atom ID.
+ */
+export class TreeIndex {
+	private detachedNodeToField: NestedMap<Major, Minor, Entry> = new Map<
+		Major,
+		Map<Minor, Entry>
+	>();
+
+	public constructor(
+		private readonly name: string,
+		private rootIdAllocator: IdAllocator<ForestRootId>,
+	) {}
+
+	public clone(): TreeIndex {
+		const clone = new TreeIndex(this.name, this.rootIdAllocator);
```

Comment:
Yes because the IDs should still be unique between all related branches, especially since they can be merged back and field keys could potentially conflict.